cli/src/tasks/migrate.ts

```diff
@@ -0,0 +1,590 @@
+    if (allDependencies[plugin]) {
+      result.push(`${plugin}@${pluginVersion}`);
+    }
+  }
+
+  await getCommandOutput('npm', ['i', ...result]);
+}
+
+async function migrateStoragePluginToPreferences() {
+  if (allDependencies['@capacitor/storage']) {
+    await getCommandOutput('npm', ['uninstall', '@capacitor/storage']);
+    await getCommandOutput('npm', [
+      'i',
+      `@capacitor/preferences@${pluginVersion}`,
+    ]);
+    logger.info('Migrated @capacitor/storage to @capacitor/preferences.');
+  }
+}
+
+function writeBreakingChanges() {
+  const breaking = [
+    '@capacitor/storage',
+    '@capacitor/camera',
+    '@capacitor/push-notifications',
+    '@capacitor/local-notifications',
+  ];
+  const broken = [];
+  for (const lib of breaking) {
+    if (allDependencies[lib]) {
+      broken.push(lib);
+    }
+  }
+  if (broken.length > 0) {
+    logger.info(
+      `IMPORTANT: Review https://next.capacitorjs.com/docs/updating/4-0#plugins for breaking changes in these plugins that you use: ${broken.join(
+        ', ',
+      )}.`,
+    );
+  }
+  if (allDependencies['@capacitor/android']) {
+    logger.info(
+      'Warning: The Android Gradle plugin was updated and it requires Java 11 to run. You may need to select this in Android Studio.',
+    );
+  }
+}
+
+function updateAndroidManifest(filename: string) {
+  const txt = readFile(filename);
+  if (!txt) {
+    return;
+  }
+
+  // AndroidManifest.xml add attribute: <activity android:exported="true"
+  if (txt.includes('<activity android:exported="')) {
+    return; // Probably already updated manually
+  }
+  const replaced = setAllStringIn(
+    txt,
+    '<activity',
+    ' ',
+    ' android:exported="true"',
+  );
+  if (txt == replaced) {
+    logger.error(`Unable to update Android Manifest. Missing <activity> tag`);
+    return;
+  }
+  writeFileSync(filename, replaced, 'utf-8');
+  logger.info(
+    `Migrated AndroidManifest.xml by adding android:exported attribute to Activity.`,
+  );
+}
+
+function updateBuildGradle(filename: string) {
+  // In build.gradle add dependencies:
+  // classpath 'com.android.tools.build:gradle:7.2.1'
+  // classpath 'com.google.gms:google-services:4.3.10'
+  const txt = readFile(filename);
+  if (!txt) {
+    return;
+  }
+  const neededDeps: { [key: string]: string } = {
+    'com.android.tools.build:gradle': '7.2.1',
+    'com.google.gms:google-services': '4.3.10',
+  };
+  let replaced = txt;
+
+  for (const dep of Object.keys(neededDeps)) {
+    if (!replaced.includes(`classpath '${dep}`)) {
+      replaced = txt.replace(
+        'dependencies {',
+        `dependencies {\n        classpath '${dep}:${neededDeps[dep]}'`,
+      );
+    } else {
+      // Update
+      replaced = setAllStringIn(
+        replaced,
+        `classpath '${dep}:`,
+        `'`,
+        neededDeps[dep],
+      );
+      logger.info(`Migrated build.gradle set ${dep} = ${neededDeps[dep]}.`);
+    }
+  }
+
+  // Replace jcenter()
```

Comment:
the upgrade guide has this step as optional since users might still have some plugin that depends on some library that is only in jcenter, so maybe it should prompt so user is aware of the change?